src/goto-instrument/loop_utils.cpp

```diff
@@ -51,6 +51,10 @@ void get_modifies_lhs(
     for(const auto &mod : local_may_alias.get(t, ptr.pointer))
     {
       const typecast_exprt typed_mod{mod, ptr.pointer.type()};
+      if(mod.id() == ID_unknown)
+      {
+        throw analysis_exceptiont("Alias analysis returned UNKNOWN!");
+      }
```

Comment:
Yes, so what?

Reply:
So, to guarantee soundness, in this context one would have to add _every visible object_ to the `modifies` set in this case (unknown == \top).

I raise an exception because I was not sure how to even do that. This is caught by the caller and processed in whatever way appropriate in the calling context. In our case, loop contract instrumentation calls this function and shows a message to the user to add an assigns clause on the loop.

Reply:
I guess it's not strictly speaking necessary for soundness. You could chose the empty set or some random junk. The abstraction of the loop would be unsound, but the step-case check would fail because we wouldn't be able to prove that you only wrote what we said.

Having said that though, surely this would be easier for the user to debug then having them try to debug the fail of the proof that the loop didn't write to the empty set of targets.

Reply:
> I guess it's not strictly speaking necessary for soundness. You could chose the empty set or some random junk. 

I think that would be unsound. Because an empty set would be an underapproximation and would lead to proving incorrect properties (that should actually fail). Consider for example:

```
a[5] = 0;
for(int i = 0; i < sz; ++i) a[i] = 1; // sz > 5
assert(a[5] == 0);
```

Suppose the alias analysis couldn't compute the write set of the loop, i.e, it returned `unknown` on the LHS `a[i]` for the statement `a[i] = 1` in the loop body.

If we consider the write set of this loop to be the empty set, we would be able to prove the postcondition, but we actually shouldn't.

Reply:
But, don't we then instrument the body with that (empty) write set and fail the step-case of the proof?

Reply:
Ah I see the confusion now.
We only instrument the body if an assigns clause was explicitly specified by the user.

Reply:
Ah I see. So that write set inference is trusted code.

Reply:
Yes exactly, all internal CBMC analyses (in this case, the alias analysis) are trusted.